### PR TITLE
feat(compiler): Add extended diagnostic to warn when the track functi…

### LINKED ISSUE #58264: Using a track function directly in the new @for loop is not supported nor compiled
### Which @angular/* package(s) are the source of the bug?

core

### Is this a regression?

No

### Description

The following code

```
@for (a of b; track method) {}
```

and having
```
method = (index, item) => { return index }
```

The method is actually not called.

It comes to a surprise because the language service will link properly and everything but it doesn't actually run the tracking method.
For someone who comes from an `*ngFor`, they will blindly switch this `trackBy` to `track` and things will start failing at runtime.

If this is the "norm" then maybe the documentation should be reflected that you have to call `method($index, a)`, but this is not obvious

### Please provide a link to a minimal reproduction of the bug

https://angular-e967ee.stackblitz.io

### Please provide the exception or error you saw

_No response_

### Please provide the environment you discovered this bug in (run `ng version`)

Angular CLI: 17.3.2
Node: 20.11.1
Package Manager: npm 8.5.5
OS: win32 x64

Angular: 17.3.2
... animations, cdk, cli, common, compiler, compiler-cli, core
... forms, google-maps, language-service, localize, material
... platform-browser, platform-browser-dynamic, platform-server
... router, ssr

Package                         Version
---------------------------------------------------------
@angular-devkit/architect       0.1703.2
@angular-devkit/build-angular   17.3.2
@angular-devkit/core            17.3.2
@angular-devkit/schematics      17.3.2
@schematics/angular             17.3.2
rxjs                            7.8.1
typescript                      5.2.2
zone.js                         0.14.4

### Anything else?

_No response_

### PR DESCRIPTION
**Pull request by eneajaho**
…on is not called on @for blocks

## PR Checklist
Please check if your PR fulfills the following requirements:

- [x] The commit message follows our guidelines: https://github.com/angular/angular/blob/main/contributing-docs/commit-message-guidelines.md
- [x] Tests for the changes have been added (for bug fixes / features)
- [ ] Docs have been added / updated (for bug fixes / features)

## PR Type
What kind of change does this PR introduce?
- [x] Feature

Fixes #58264

## What is the current behavior?
Passing just the function name to the track function basically will make the for loop recreate the list everytime. The function need to be called.

## What is the new behavior?
A template diagnostic will warn the developer that they need to call the function in the track part of a @for block for desired performance outcome.

## Does this PR introduce a breaking change?

- [x] No

### CODE REVIEW

**Comment by eneajaho** [comment]
TODOs:
- I need to add documentation page for the new warning code 
- I need to add more tests for edge cases (open for ideas)

**Comment by JeanMeche** [comment]
#59356

File: packages/compiler-cli/src/ngtsc/typecheck/extended/checks/uninvoked_track_function/index.ts [hunk 1]
```diff
@@ -0,0 +1,80 @@
...
+ */
+class UninvokedTrackFunctionCheck extends TemplateCheckWithVisitor<ErrorCode.UNINVOKED_TRACK_FUNCTION> {
+  override code = ErrorCode.UNINVOKED_TRACK_FUNCTION as const;
+
+  override visitNode(
+    ctx: TemplateContext<ErrorCode.UNINVOKED_TRACK_FUNCTION>,
+    component: ts.ClassDeclaration,
+    node: TmplAstNode | AST,
+  ): NgTemplateDiagnostic<ErrorCode.UNINVOKED_TRACK_FUNCTION>[] {
+    if (node instanceof TmplAstForLoopBlock) {
```
Comment (JeanMeche) [inline packages/compiler-cli/src/ngtsc/typecheck/extended/checks/uninvoked_track_function/index.ts:36]: We could inverse the condition and return early.

File: packages/compiler-cli/src/ngtsc/typecheck/extended/test/checks/uninvoked_track_function/uninvoked_track_function.spec.ts [hunk 2]
```diff
@@ -0,0 +1,96 @@
...
+        `
+          @for (item of items; track item.name) {}
+        `,
+        `trackByName(item) { return item.name; }`,
+      );
+      expect(diags.length).toBe(0);
+    });
+
+    it('should not produce a warning when the function is invoked', () => {
+      const diags = setupTestComponent(
```
Comment (JeanMeche) [inline packages/compiler-cli/src/ngtsc/typecheck/extended/test/checks/uninvoked_track_function/uninvoked_track_function.spec.ts:56]: Can we add a case to cover getters ? As in a getter should not trigger the diagnostic.

File: packages/compiler-cli/src/ngtsc/typecheck/extended/checks/uninvoked_track_function/index.ts [hunk 3]
```diff
@@ -0,0 +1,81 @@
...
+      !(node.trackBy.ast instanceof SafePropertyRead)
+    ) {
+      // If the expression is not a property read, skip it.
+      return [];
+    }
+
+    const symbol = ctx.templateTypeChecker.getSymbolOfNode(node.trackBy.ast, component);
+
+    if (symbol !== null && symbol.kind === SymbolKind.Expression) {
+      if (symbol.tsType.getCallSignatures()?.length > 0) {
```
Comment (JeanMeche) [inline packages/compiler-cli/src/ngtsc/typecheck/extended/checks/uninvoked_track_function/index.ts:56]:
```suggestion
    if (symbol?.kind === SymbolKind.Expression && symbol.tsType.getCallSignatures()?.length > 0) {
```

nit: We could make it a single condition.

**Review by JeanMeche — APPROVED** [review]
Beside the small nit.

File: packages/compiler-cli/src/ngtsc/typecheck/extended/test/checks/uninvoked_track_function/uninvoked_track_function.spec.ts (hunk 2 above)
Comment (alfaproject) [reply packages/compiler-cli/src/ngtsc/typecheck/extended/test/checks/uninvoked_track_function/uninvoked_track_function.spec.ts:57]: Isn't a getter (from the class, not from an item) also going to recreate the list every time? Sounds like something we'd like to warn about as well, right?

Comment (JeanMeche) [reply packages/compiler-cli/src/ngtsc/typecheck/extended/test/checks/uninvoked_track_function/uninvoked_track_function.spec.ts:57]: I was talking about the track function being a getter. 
But a getter is probably also a bad track function as it doesn't take any argument. We should maybe has a specific error about it.

**Comment by JeanMeche** [comment]
This doesn't pass the presubmit, it correctly flagged un-invoked functions.

File: packages/compiler-cli/src/ngtsc/typecheck/extended/checks/uninvoked_track_function/index.ts [hunk 4]
```diff
@@ -0,0 +1,83 @@
...
+    if (
+      symbol !== null &&
+      symbol.kind === SymbolKind.Expression &&
+      symbol.tsType.getCallSignatures()?.length > 0
+    ) {
+      const fullExpressionText = generateStringFromExpression(
+        node.trackBy.ast,
+        node.trackBy.source || '',
+      );
+      const errorString = `The track function in the @for block should be invoked: ${fullExpressionText}(${node.item.name})`;
```
Comment (JeanMeche) [inline packages/compiler-cli/src/ngtsc/typecheck/extended/checks/uninvoked_track_function/index.ts:64]: Passing the item as argument might be misleading, it depends on the shape of the tracking function. Should we maybe go with something like `${fullExpressionText}(/* arguments*/)`

**Review by devversion — APPROVED** [review]
LGTM from my side, adding @crisbeto in case he has other thoughts.

**Review by crisbeto — APPROVED** [review]

**Comment by crisbeto** [comment]
FYI this will require a TGP, because extended diagnostics can break existing targets.

**Comment by JeanMeche** [comment]
I'll take care of the cleanup.

File: packages/compiler-cli/src/ngtsc/typecheck/extended/test/checks/uninvoked_track_function/uninvoked_track_function.spec.ts [hunk 5]
```diff
@@ -0,0 +1,97 @@
...
+        `,
+        `trackByName(item) { return item.name; }`,
+      );
+
+      expect(diags.length).toBe(0);
+    });
+  });
+});
+
+function setupTestComponent(template: string, classField: string) {
```
Comment (kirjs) [inline packages/compiler-cli/src/ngtsc/typecheck/extended/test/checks/uninvoked_track_function/uninvoked_track_function.spec.ts:69]: nit/optional, maybe call it `diagnoseTestComponent` or something similare since in doesn't just setup the component, but also runs the diagnostics

File: packages/compiler-cli/src/ngtsc/typecheck/extended/test/checks/uninvoked_track_function/uninvoked_track_function.spec.ts [hunk 6]
```diff
@@ -0,0 +1,97 @@
...
+runInEachFileSystem(() => {
+  describe('UninvokedTrackFunctionCheck', () => {
+    it('binds the error code to its extended template diagnostic name', () => {
+      expect(uninvokedTrackFunctionCheckFactory.code).toBe(ErrorCode.UNINVOKED_TRACK_FUNCTION);
+      expect(uninvokedTrackFunctionCheckFactory.name).toBe(
+        ExtendedTemplateDiagnosticName.UNINVOKED_TRACK_FUNCTION,
+      );
+    });
+
+    it('should produce a diagnostic when a track function in a @for block is not invoked', () => {
```
Comment (kirjs) [inline packages/compiler-cli/src/ngtsc/typecheck/extended/test/checks/uninvoked_track_function/uninvoked_track_function.spec.ts:28]: nit: can we also have a test to verify that good old ` track item.name` continues working?

**Review by kirjs — APPROVED** [review]
Looks great, thanks for adding this

Reviewed-for: public-api

File: packages/compiler-cli/src/ngtsc/typecheck/extended/test/checks/uninvoked_track_function/uninvoked_track_function.spec.ts [hunk 7]
```diff
@@ -0,0 +1,108 @@
...
+      expect(diags.length).toBe(1);
+      expect(diags[0].category).toBe(ts.DiagnosticCategory.Warning);
+      expect(diags[0].code).toBe(ngErrorCode(ErrorCode.UNINVOKED_TRACK_FUNCTION));
+      expect(getSourceCodeForDiagnostic(diags[0])).toBe(
+        `@for (item of items; track trackByName) {}`,
+      );
+      expect(diags[0].messageText).toBe(generateDiagnosticText('trackByName'));
+    });
+
+    it('should not produce a warning when track is set to a getter', () => {
```
Comment (alfaproject) [inline packages/compiler-cli/src/ngtsc/typecheck/extended/test/checks/uninvoked_track_function/uninvoked_track_function.spec.ts:45]: I actually hope you guys will reevaluate this decision. It should 100% produce a warning

This MR is already good as is, but this test tickles my OCD d:

**Comment by JeanMeche** [comment]
[TGP](https://fusion2.corp.google.com/presubmit/739674169/OCL:739674169:BASE:745172107:1744128901078:5b4005c6/targets) is  green.

**Comment by atscott** [comment]
This PR was merged into the repository by commit 7a971766dc59691dc68da9439e180a6c4d7b17d8.

The changes were merged into the following branches: main
